docs/types/literal-suffixes.rst

```diff
@@ -0,0 +1,269 @@
+    avoid explicit conversions when the return type is signed.
+
+.. note::
+    40-digit literals prefixed with ``0x`` such as, for example, ``0xdCad3a6d3569DF655070DEd06cb7A1b2Ccd1D3AF``
+    always represent ``address`` literals in the language.
+    To invoke a suffix accepting ``bytes20`` you must use one of the other literal kinds implicitly
+    convertible to ``bytes20``, e.g., a hexadecimal string literal
+    (``hex"dCad3a6d3569DF655070DEd06cb7A1b2Ccd1D3AF"``).
+
+.. note::
+    Suffix functions accepting ``address payable`` are not allowed since address literals are never payable.
+
+Suffix functions may not accept or return reference types with ``storage`` or ``calldata`` locations.
+
+.. index:: ! fractional decomposition
+.. _fractional_decomposition:
+
+Fractional Decomposition
+------------------------
+
+To allow defining suffixes that work with fractional literals, like ``1.23``, the language allows
+a special form of a suffix definition.
+Such a suffix can be considered a more general form of a suffix taking a single integer argument.
+
+A single-parameter suffix can be applied only to those rational number literals which represent
+integers.
+Let us consider the following suffix definition:
+
+.. code-block:: solidity
+
+    function kg(uint grams) pure suffix returns (uint) {
+        return 1000 * grams;
+    }
+
+The ``kg`` suffix can receive integer values like ``123 kg``, ``1.23e2 kg``, or ``12300e-2 kg``.
+However, invoking such a suffix with a fractional number (e.g., ``1.23 kg``) triggers an error.
+We can fix that by adding an *exponent* parameter:
+
+.. code-block:: solidity
+
+    function kg(uint mantissa, uint exponent) pure suffix returns (uint) {
+        return 1000 * mantissa / 10**exponent;
+    }
+
+When defined this way, the suffix can handle all the literals it could previously, while ``1.23 kg``
+also becomes a valid expression, equivalent to ``kg(123, 2)``.
+
+More generally, the argument of such a suffix call is decomposed into two integer values (``mantissa``
+and ``exponent``), such that:
+
+#. ``mantissa * 10**-exponent`` is equal to the value of the literal.
+#. ``exponent`` is the smallest possible non-negative integer value satisfying the equation.
+
+The two rules provide unambiguous decomposition in all cases.
+For example:
+
+- ``123000`` is decomposed into ``123000 * 10**-0`` (i.e. ``123000`` for ``mantissa`` and ``0`` for ``exponent``).
+  Not ``123 * 10**3`` or ``123000000 * 10**-3``.
+
+  In general, when the suffix is invoked on an integer, ``mantissa`` is always equal to that integer
+  and ``exponent`` is ``0``.
+- ``1.23`` is decomposed into ``123 * 10**-2``, not ``1.23 * 10**-0`` or ``123000 * 10**-5``.
+
+  In general, when the suffix is invoked on a fractional number, ``exponent`` is the negation of
+  the lowest negative power of ``10`` that, when multiplied by the literal, produces an integer value.
```

Comment:
Assuming the literal ``1.23``
The lowest power of 10, s.t. it multiplied by the literal is an integer is ``10**2``, since ``1.23 * 10**2`` is integer.
So exponent is ``-100`` :-)? That's the negation of that power of two... the negation of merely the exponent of that power of two would be ``-2``... still not exactly right...

Reply:
What would you suggest though? I know it may not be a very precise phrasing, but I had a problem with explaining concisely. It's generally a confusing thing.

Reply:
That's no reason for formulating it wrongly, though, is it :-)? You either have to refer to multiplying the mantissa with 10 to the power of the negative exponent to get the literal (so opposite to what you say in the text here) - or the other way round talk about the (positive) exponent of the lowest positive power of 10 (without anything being negated), s.t. multiplying it by the literal you get an integer.
The current text seems to mix the two and thereby gets all signs wrong.

Reply:
By the way, on a related note, what do you think about the error message that you get when you try to use a signed type for exponent:
```c++
"The exponent parameter of a literal suffix function must be unsigned. "
"Exponent is always either zero or a negative power of 10 but the parameter represents its absolute value."
```

I could just stop at saying that it must be unsigned, but, since this is counter-intuitive (it's really a negated exponent), I thought it's best if the message at least mentions why, so that there's a smaller chance people will incorrectly assume we use a positive power of 10 just because the compiler does not let them use a signed type. Still, I'm not very satisfied with that wording and I suspect you'd have some reservations about it too (it conflates the exponent of the power of ten with a power of ten the same way my docs did). Still, it's hard to come up with something concise here.